File: src/main/services/analysis/ProcessLinker.ts
/**
 * ProcessLinker service - Links subagent processes to AI chunks.
 *
 * Uses a two-tier linking strategy:
 * 1. Primary: parentTaskId matching - Links subagents to chunks containing the Task tool call
 *    that spawned them. This is reliable even when the response is still in progress.
 * 2. Fallback: Timing-based - For orphaned subagents without parentTaskId, falls back to
 *    checking if the subagent's startTime falls within the chunk's time range.
 */

import { type EnhancedAIChunk, type Process } from '@main/types';

/**
 * Link processes to a single AI chunk.
 *
 * Uses a two-tier linking strategy:
 * 1. Primary: parentTaskId matching - Links subagents to chunks containing the Task tool call
 *    that spawned them. This is reliable even when the response is still in progress.
 * 2. Fallback: Timing-based - For orphaned subagents without parentTaskId, falls back to
 *    checking if the subagent's startTime falls within the chunk's time range.
 */
export function linkProcessesToAIChunk(chunk: EnhancedAIChunk, subagents: Process[]): void {
  // Build set of Task tool IDs from this chunk's responses
  const chunkTaskIds = new Set<string>();
  for (const response of chunk.responses) {
    for (const toolCall of response.toolCalls) {
      if (toolCall.isTask) {
        chunkTaskIds.add(toolCall.id);
      }
    }
  }

  // Track which subagents have been linked
  const linkedSubagentIds = new Set<string>();

  // Primary linking: Match subagents to Task calls by parentTaskId
  for (const subagent of subagents) {
    if (subagent.parentTaskId && chunkTaskIds.has(subagent.parentTaskId)) {
      chunk.processes.push(subagent);
      linkedSubagentIds.add(subagent.id);
    }
  }

  // Fallback linking: For orphaned subagents, use timing-based matching
  // This handles edge cases where parentTaskId might not be set
  for (const subagent of subagents) {
    if (linkedSubagentIds.has(subagent.id)) {
      continue; // Already linked via parentTaskId
    }

    // Only use timing fallback if subagent has no parentTaskId
    // (If it has parentTaskId but didn't match, it belongs to a different chunk)
    if (!subagent.parentTaskId) {
      if (subagent.startTime >= chunk.startTime && subagent.startTime <= chunk.endTime) {
        chunk.processes.push(subagent);
      }
    }
  }

  chunk.processes.sort((a, b) => a.startTime.getTime() - b.startTime.getTime());
}

File: src-tauri/src/analysis/process_linker.rs
/// Link subagent processes to AI chunks.
/// Two-tier: parentTaskId matching (primary), timing-based (fallback).

use crate::types::chunks::Process;
use crate::types::messages::ParsedMessage;

/// Link processes to an AI chunk and return the linked ones.
pub fn link_processes_to_ai_chunk(
    chunk_responses: &[ParsedMessage],
    chunk_start: &str,
    chunk_end: &str,
    subagents: &[Process],
) -> Vec<Process> {
    // Build set of Task tool IDs from this chunk's responses
    let mut chunk_task_ids = std::collections::HashSet::new();
    for resp in chunk_responses {
        for tc in &resp.tool_calls {
            if tc.is_task {
                chunk_task_ids.insert(tc.id.as_str());
            }
        }
    }

    let mut linked = Vec::new();
    let mut linked_ids = std::collections::HashSet::new();

    // Primary: parentTaskId matching
    for sub in subagents {
        if let Some(ref ptid) = sub.parent_task_id {
            if chunk_task_ids.contains(ptid.as_str()) {
                linked.push(sub.clone());
                linked_ids.insert(sub.id.as_str());
            }
        }
    }

    // Fallback: timing-based for orphaned subagents (no parentTaskId)
    for sub in subagents {
        if linked_ids.contains(sub.id.as_str()) {
            continue;
        }
        if sub.parent_task_id.is_some() {
            continue; // Has parentTaskId but didn't match — belongs to different chunk
        }
        if sub.start_time.as_str() >= chunk_start && sub.start_time.as_str() <= chunk_end {
            linked.push(sub.clone());
        }
    }

    linked.sort_by(|a, b| a.start_time.cmp(&b.start_time));
    linked
}
